File: backend/sdk/tools/item_tools.py
"""
Item management tools for TRPG gameplay.

Contains tools for item creation and management:
- persist_item: Create item templates (used by sub-agents via Task tool)

Uses ItemService for filesystem-first item template management.
When add_to_inventory is True, items are also added to player inventory
via PlayerFacade (which syncs to DB for frontend display).

Supports batch creation of multiple items in a single call.
"""

import logging
from typing import Any

from claude_agent_sdk import tool
from services.facades import PlayerFacade
from services.item_service import ItemService

from sdk.config.subagent_tool_definitions import PersistItemInput
from sdk.loaders import get_tool_description, is_tool_enabled
from sdk.tools.context import ToolContext

logger = logging.getLogger("GameplayTools.Item")


def create_item_tools(ctx: ToolContext) -> list:
    """
    Create item management tools.

    Args:
        ctx: Unified tool context containing agent info and dependencies

    Returns:
        List of item tool functions
    """
    tools = []

    # Get required dependencies from context
    world_name = ctx.require_world_name()

    # Get optional db/world_id for DB sync (may be None for sub-agents)
    db = ctx.db
    world_id = ctx.world_id

    # Create PlayerFacade for FS-first with DB sync (if available)
    player_facade = PlayerFacade(world_name, db=db, world_id=world_id) if db and world_id else None

    # ==========================================================================
    # persist_item tool - Create item templates in items/ directory
    # ==========================================================================
    if is_tool_enabled("persist_item", default=True):
        persist_item_description = get_tool_description(
            "persist_item",
            agent_name="Item Designer",
            group_name=ctx.group_name,
            default="Persist one or more item designs to the game world. "
            "Creates item templates in filesystem (items/[item_id].yaml).",
        )

        @tool(
            "persist_item",
            persist_item_description,
            PersistItemInput.model_json_schema(),
        )
        async def persist_item_tool(args: dict[str, Any]):
            """Persist item designs to the game world.

            Supports creating multiple items in a single call.
            Items are created here by Item Designer sub-agent before they can be
            added to player inventory via change_stat.
            """
            validated = PersistItemInput(**args)

            logger.info(f"📦 persist_item: {len(validated.items)} item(s)")

            created_items = []
            skipped_items = []
            inventory_added = []

            try:
                # Process each item
                for item in validated.items:
                    # Check if item already exists
                    existing = ItemService.load_item_template(world_name, item.item_id)
                    if existing:
                        skipped_items.append(item.item_id)
                        logger.warning(f"⚠️ Item '{item.item_id}' already exists, skipping")
                        continue

                    # Save the new item template with all new optional fields
                    ItemService.save_item_template(
                        world_name,
                        item_id=item.item_id,
                        name=item.name,
                        description=item.description,
                        properties=item.properties,
                        # NEW fields for world-agnostic item system
                        category=item.category,
                        tags=item.tags,
                        rarity=item.rarity,
                        icon=item.icon,
                        stacking=item.stacking.model_dump() if item.stacking else None,
                        equippable=item.equippable.model_dump() if item.equippable else None,
                        usable=item.usable.model_dump() if item.usable else None,
                    )
                    created_items.append(item)
                    logger.info(f"✅ Created item template: {item.item_id}")

                # Add to inventory if requested (used during onboarding for starting items)
                if validated.add_to_inventory and created_items:
                    if player_facade:
                        # Use PlayerFacade for FS-first with DB sync
                        for item in created_items:
                            await player_facade.add_item(
                                item_id=item.item_id,
                                name=item.name,
                                quantity=item.quantity,
                                description=item.description,
                                properties=item.properties,
                            )
                            inventory_added.append(f"{item.quantity}x {item.name}")
                            logger.info(f"✅ Added {item.quantity}x {item.item_id} to inventory (with DB sync)")
                    else:
                        # Fallback: FS-only (no DB sync available)
                        from services.player_service import PlayerService

                        player_state = PlayerService.load_player_state(world_name)
                        if player_state:
                            for item in created_items:
                                inventory_entry = {
                                    "item_id": item.item_id,
                                    "quantity": item.quantity,
                                }
                                player_state.inventory.append(inventory_entry)
                                inventory_added.append(f"{item.quantity}x {item.name}")
                                logger.info(f"✅ Added {item.quantity}x {item.item_id} to inventory (FS-only)")
                            PlayerService.save_player_state(world_name, player_state)
                        else:
                            logger.warning("Could not load player state to add items to inventory")

                # Build response
                response_parts = []

                if created_items:
                    response_parts.append(f"**Created {len(created_items)} item(s):**")
                    for item in created_items:
                        item_line = f"- `{item.item_id}`: {item.name}"
                        if item.properties:
                            props = ", ".join(f"{k}={v}" for k, v in item.properties.items())
                            item_line += f" ({props})"
                        response_parts.append(item_line)

                if skipped_items:
                    response_parts.append(
                        f"\n**Skipped {len(skipped_items)} (already exist):** {', '.join(skipped_items)}"
                    )

                if inventory_added:
                    response_parts.append(f"\n**Added to inventory:** {', '.join(inventory_added)}")
                elif validated.add_to_inventory and not created_items:
                    response_parts.append("\n⚠️ No items added to inventory (all items already existed)")
                elif not validated.add_to_inventory and created_items:
                    response_parts.append("\nAction Manager can add these items to inventory via change_stat.")

                return {"content": [{"type": "text", "text": "\n".join(response_parts)}]}

            except Exception as e:
                logger.error(f"persist_item error: {e}", exc_info=True)
                return {
                    "content": [{"type": "text", "text": f"Error creating items: {e}"}],
                    "is_error": True,
                }

        tools.append(persist_item_tool)

    return tools

File: backend/src/sdk/handlers/item-tools.ts
import type { ItemService } from '../../services/item-service'
import type { PlayerService } from '../../services/player-service'
import { persistItemTool } from '../tools/subagent'
import type { ItemDefinitionInput } from '../tools/subagent'
import { resolveTool } from '../tools/registry'
import { getLogger } from '../../infrastructure/logging/logger'
import type { PlayerMutationsPort } from './ports'
import { tool, requireWorldName, toolError, toolSuccess, type SdkTool, type ToolContext } from './context'

/**
 * `persist_item` — the Item Designer sub-agent's callback. Batched on purpose:
 * onboarding creates a dozen starting items and one call each would be a dozen
 * round trips. An existing id is skipped and reported, never overwritten — the
 * designer runs repeatedly and must not rewrite an author's edits.
 */

const logger = getLogger('GameplayTools.Item')

/** Drop `null`-valued keys, so an absent field is absent rather than explicit. */
function stripNulls(value: Record<string, unknown>): Record<string, unknown> {
  return Object.fromEntries(Object.entries(value).filter(([, v]) => v !== null))
}

export interface ItemToolDeps {
  items: ItemService
  players: PlayerService
  /**
   * Present during gameplay, absent for a bare sub-agent context. Without it
   * the inventory append is filesystem-only, which onboarding's item designer
   * needs: it runs before the world has a room to sync against.
   */
  mutations?: PlayerMutationsPort
}

export function createItemTools(ctx: ToolContext, deps: ItemToolDeps): SdkTool[] {
  const worldName = requireWorldName(ctx)

  const def = resolveTool(persistItemTool.name, ctx.groupName)
  if (!def) return []

  return [
    tool(
      persistItemTool.name,
      def.description,
      persistItemTool.inputSchema,
      async (args) => {
        const created: ItemDefinitionInput[] = []
        const skipped: string[] = []
        const inventoryAdded: string[] = []

        try {
          for (const item of args.items) {
            if (deps.items.loadItemTemplate(worldName, item.item_id)) {
              skipped.push(item.item_id)
              logger.warning(`Item '${item.item_id}' already exists, skipping`)
              continue
            }

            deps.items.saveItemTemplate(worldName, {
              itemId: item.item_id,
              name: item.name,
              description: item.description,
              properties: item.properties,
              category: item.category ?? null,
              tags: item.tags ?? null,
              rarity: item.rarity ?? null,
              icon: item.icon ?? null,
              stacking: item.stacking ?? null,
              // `Equippable`'s optional fields are `T | undefined`, never
              // `T | null`: the item service tests them for absence, so a
              // written-out `accepts_as: null` would read as present.
              equippable: item.equippable ? stripNulls(item.equippable) : null,
              usable: item.usable ?? null,
            })
            created.push(item)
          }

          if (args.add_to_inventory && created.length > 0) {
            if (deps.mutations) {
              for (const item of created) {
                deps.mutations.addItem(worldName, {
                  itemId: item.item_id,
                  name: item.name,
                  quantity: item.quantity,
                  description: item.description,
                  properties: item.properties,
                })
                inventoryAdded.push(`${item.quantity}x ${item.name}`)
              }
            } else {
              const state = deps.players.loadPlayerState(worldName)
              if (state) {
                for (const item of created) {
                  // Reference form only — the template just created carries the
                  // name and description.
                  state.inventory.push({ item_id: item.item_id, quantity: item.quantity })
                  inventoryAdded.push(`${item.quantity}x ${item.name}`)
                }
                deps.players.savePlayerState(worldName, state)
              } else {
                logger.warning('Could not load player state to add items to inventory')
              }
            }
          }

          const parts: string[] = []

          if (created.length > 0) {
            parts.push(`**Created ${created.length} item(s):**`)
            for (const item of created) {
              let line = `- \`${item.item_id}\`: ${item.name}`
              if (Object.keys(item.properties).length > 0) {
                const props = Object.entries(item.properties)
                  .map(([key, value]) => `${key}=${String(value)}`)
                  .join(', ')
                line += ` (${props})`
              }
              parts.push(line)
            }
          }

          if (skipped.length > 0) {
            parts.push(`\n**Skipped ${skipped.length} (already exist):** ${skipped.join(', ')}`)
          }

          if (inventoryAdded.length > 0) {
            parts.push(`\n**Added to inventory:** ${inventoryAdded.join(', ')}`)
          } else if (args.add_to_inventory && created.length === 0) {
            parts.push('\n⚠️ No items added to inventory (all items already existed)')
          } else if (!args.add_to_inventory && created.length > 0) {
            parts.push('\nAction Manager can add these items to inventory via change_stat.')
          }

          return toolSuccess(parts.join('\n'))
        } catch (error) {
          logger.error(`persist_item error: ${String(error)}`)
          return toolError(`Error creating items: ${String(error)}`)
        }
      },
    ),
  ]
}
